### PR TITLE
fix: redirect bug on dataset view when referrer is external

### PR DESCRIPTION
**Pull request by davidverdu**
fixes: https://github.com/IQSS/ondemand-loop/issues/239

### CODE REVIEW

File: application/app/controllers/dataverse/datasets_controller.rb [hunk 1]
```diff
@@ -107,4 +107,13 @@ def search_files_page
       return
     end
   end
+
+  def redirect_to_safe_referer(fallback: root_path, **options)
```
Comment (abujeda) [inline application/app/controllers/dataverse/datasets_controller.rb:111]: not sure about the name. What about `redirect_back_to_app` or similar?
safe is a bit misleading.

File: application/app/controllers/dataverse/datasets_controller.rb [hunk 2]
```diff
@@ -107,4 +107,13 @@ def search_files_page
...
+    referer = request.referer
+    if referer.present? && URI.parse(referer).host == request.host
```
[the remark this answers is not in the document]
Comment (abujeda) [reply application/app/controllers/dataverse/datasets_controller.rb:113]: maybe check if the referrer path starts with /pun/sys/loop?
I think in Rake, this is passed with `request.script_name`.

This needs to be tested to see how to get the mount point from Passenger. As it would be deployed under a different URL

File: application/app/controllers/dataverse/datasets_controller.rb [hunk 3]
```diff
@@ -107,4 +107,27 @@ def search_files_page
...
+
+  def redirect_back_to_app(fallback: root_path, **options)
+    if internal_referer?
+      redirect_back(fallback_location: fallback, **options)
+    else
+      redirect_to(fallback, **options)
+    end
+  end
+
+  def internal_referer?
```
Comment (abujeda) [inline application/app/controllers/dataverse/datasets_controller.rb:119]: To make the name aligned to the behaviour, I would rename it to `redirect_back?`

File: application/test/controllers/dataverse/datasets_controller_test.rb [hunk 4]
```diff
@@ -77,6 +103,32 @@ def files_incomplete_no_data_file_json
...
+  test "should redirect to internal referrer after raising exception with script name" do
+    Dataverse::DatasetService.any_instance.stubs(:find_dataset_version_by_persistent_id).raises("error")
+    Dataverse::DatasetService.any_instance.stubs(:search_dataset_files_by_persistent_id).returns(nil)
+    internal_referer = "/pun/sys/loop" + view_dataverse_landing_path.to_s
+    get view_dataverse_dataset_url(@new_id, "random_id"), headers: { "HTTP_REFERER" => internal_referer }, env: { "SCRIPT_NAME" => "/pun/sys/loop" }
+    assert_redirected_to internal_referer
+    assert_equal "Dataverse service error. Dataverse: https://#{@new_id} persistentId: random_id", flash[:alert]
+  end
+
+  test "should redirect to fallback url after raising exception coming from external referrer" do
```
Comment (abujeda) [inline application/test/controllers/dataverse/datasets_controller_test.rb:124]: fallback is an implementation details. This test asserts that redirects to the root path.

File: application/test/controllers/dataverse/datasets_controller_test.rb [hunk 5]
```diff
@@ -85,6 +137,32 @@ def files_incomplete_no_data_file_json
...
+  test "should redirect to internal referrer after raising Unauthorized exception with script name" do
+    Dataverse::DatasetService.any_instance.stubs(:find_dataset_version_by_persistent_id).raises(Dataverse::DatasetService::UnauthorizedException)
+    Dataverse::DatasetService.any_instance.stubs(:search_dataset_files_by_persistent_id).raises(Dataverse::DatasetService::UnauthorizedException)
+    internal_referer = "/pun/sys/loop" + view_dataverse_landing_path.to_s
+    get view_dataverse_dataset_url(@new_id, "random_id"), headers: { "HTTP_REFERER" => internal_referer }, env: { "SCRIPT_NAME" => "/pun/sys/loop" }
+    assert_redirected_to internal_referer
+    assert_equal "Dataset requires authorization. Dataverse: https://#{@new_id} persistentId: random_id", flash[:alert]
+  end
+
+  test "should redirect to fallback url after raising Unauthorized exception coming from external referrer" do
```
Comment (abujeda) [inline application/test/controllers/dataverse/datasets_controller_test.rb:158]: same here

File: application/test/controllers/dataverse/datasets_controller_test.rb [hunk 6]
```diff
@@ -85,6 +137,32 @@ def files_incomplete_no_data_file_json
...
+    Dataverse::DatasetService.any_instance.stubs(:find_dataset_version_by_persistent_id).raises(Dataverse::DatasetService::UnauthorizedException)
+    Dataverse::DatasetService.any_instance.stubs(:search_dataset_files_by_persistent_id).raises(Dataverse::DatasetService::UnauthorizedException)
+    get view_dataverse_dataset_url(@new_id, "random_id"), headers: { HTTP_REFERER: "http://external.com/another/page"}, env: { "SCRIPT_NAME" => "/pun/sys/loop" }
+    assert_redirected_to root_path
+    assert_equal "Dataset requires authorization. Dataverse: https://#{@new_id} persistentId: random_id", flash[:alert]
+  end
+
```
Comment (abujeda) [inline application/test/controllers/dataverse/datasets_controller_test.rb:165]: missing a test for when script_name is nil

File: application/app/controllers/dataverse/datasets_controller.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (davidverdu) [reply application/app/controllers/dataverse/datasets_controller.rb]: implemented with `request.script_name`

File: application/app/controllers/dataverse/datasets_controller.rb [hunk 7]
```diff
@@ -107,4 +107,27 @@ def search_files_page
...
+
+    begin
+      referer_uri = URI.parse(request.referer)
+    rescue URI::InvalidURIError
+      return false
+    end
+
+    return true if referer_uri.host == request.host
+    script_name = ::Configuration.loop_path
+    return referer_uri.path.start_with?(script_name)
```
[the remark this answers is not in the document]
Comment (davidverdu) [reply application/app/controllers/dataverse/datasets_controller.rb]: I removed the config variable

**Review by abujeda — APPROVED** [review]
